optuna_dashboard/ts/components/GraphParetoFront.tsx

```diff
@@ -165,6 +165,58 @@ const makeMarker = (
   }
 }
 
+const getIsDominatedTrialND = (normalizedValues: number[][]) => {
+  // Fallback for straight-forward pareto front algorithm (O(N^2) complexity).
+  const dominatedTrials: boolean[] = []
+  normalizedValues.forEach((values0: number[]) => {
+    const dominated = normalizedValues.some((values1: number[]) => {
+      if (values0.every((value0: number, k: number) => values1[k] === value0)) {
+        return false
+      }
+      return values0.every((value0: number, k: number) => values1[k] <= value0)
+    })
+    dominatedTrials.push(dominated)
+  })
+  return dominatedTrials
+}
+
+const getIsDominatedTrial2D = (normalizedValues: number[][]) => {
+  // Fast pareto front algorithm (O(N log N) complexity).
+  const sorted = normalizedValues
+    .map((values, i) => [values[0], values[1], i])
+    .sort()
+  let minValue1 = sorted[0][1]
+  const dominatedTrials: boolean[] = new Array(normalizedValues.length).fill(
+    true
+  )
+
+  sorted.forEach((values) => {
+    if (values[1] <= minValue1) {
```

Comment:
This condition is buggy when the values between trials are the same.

Reply:
In that case, both trials should be considered undominated.

Reply:
Now given a sequence of 2D vectors $\\{(a_n, b_n)\\}_{n=1}^N$ such that $a_1 \leq a_2 \leq \dots \leq a_N$ and $b_i \leq b_j$ for all $\\{(i, j) \in [N] \times [N] \mid a_i = a_j, i < j\\}$ where $[N] \coloneqq \\{1,2,\dots,N\\}$, the condition here is a bit strange when the equality holds.

For example, $a_i \leq a_j$ holds from the assumption and we now assume that $b_i = b_j$.
As mentioned above, let's assume that $a_i = a_j$ and $b_i = b_j$ do not dominate each other.
However, $a_i < a_j$ and $\min_{n \in [i]} b_i = b_j$ could still hold.
It means that there exists $(a_i, b_i)$ such that $a_i < a_j$ and $b_i \leq b_j$, which weakly dominates $(a_j, b_j)$.

Maybe I am wrong, if so, correct me please!

Reply:
Quick solution to this is the following:

```ts
const normalizedValues = [[0, 1], [1, 1], [2, 1]]

const getIsDominatedTrial2D = (normalizedValues: number[][]) => {
    const sortedValues = normalizedValues
      .map((values, i) => [values[0], values[1], i])
      .sort((a, b) => a[0] !== b[0] ? a[0] - b[0] : a[1] - b[1])
    let minValue1 = sortedValues[0][1]
    let latestParetoSolution = sortedValues[0].slice(0, 2)
    const dominatedTrials: boolean[] = new Array(normalizedValues.length).fill(
      true
    )
    sortedValues.forEach((values) => {
      if (values[1] < minValue1) {
        dominatedTrials[values[2]] = false
        minValue1 = values[1]
        latestParetoSolution = values.slice(0, 2)
      }else if (latestParetoSolution.every((v, i) => v === values[i])) {
        dominatedTrials[values[2]] = false
      }
    })
    return dominatedTrials
}

console.log(getIsDominatedTrial2D(normalizedValues))
```

Reply:
See https://github.com/optuna/optuna/blob/master/optuna/study/_multi_objective.py#L101. Optuna uses `weakly dominates` for comparison. It seems that @nabenabe0928 is correct.

Reply:
Thank you for pointing out. My statement is incorrect and @nabenabe0928's one is correct, I think.

Reply:
The new modification fails with the following case. 

```tsc
const normalizedValues = [[0, 1], [1, 2], [2, 1], [3, 0], [3, 0]]
```

We get:

```
[ false, true, false, false, false ]
```

But we would like to have:

```
[ false, true, true, false, false ]
```